library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,883 @@
+    ///  patches/
+    ///    <patch_number>/
+    ///      dlc.vmcode
+    ///    <patch_number>/
+    ///      dlc.vmcode
+    root_dir: PathBuf,
+
+    /// Metadata about the patches we have downloaded that is persisted to disk.
+    patches_state: PatchesState,
+}
+
+impl PatchManager {
+    /// Creates a new PatchManager with the given root directory. This directory is
+    /// assumed to exist. The PatchManager will use this directory to store its
+    /// state and patch binaries.
+    pub fn with_root_dir(root_dir: PathBuf) -> Self {
+        let patches_state = Self::load_patches_state(&root_dir).unwrap_or_default();
+
+        Self {
+            root_dir,
+            patches_state,
+        }
+    }
+
+    fn load_patches_state(root_dir: &Path) -> Option<PatchesState> {
+        let path = root_dir.join(PATCHES_STATE_FILE_NAME);
+        match disk_io::read(&path) {
+            Ok(maybe_state) => maybe_state,
+            Err(e) => {
+                error!(
+                    "Failed to load patches state from {}: {}",
+                    path.display(),
+                    e
+                );
+                None
+            }
+        }
+    }
+
+    fn save_patches_state(&self) -> Result<()> {
+        let path = self.root_dir.join(PATCHES_STATE_FILE_NAME);
+        disk_io::write(&self.patches_state, &path)
+    }
+
+    /// The directory where all patch artifacts are stored.
+    fn patches_dir(&self) -> PathBuf {
+        self.root_dir.join(PATCHES_DIR_NAME)
+    }
+
+    /// The directory where artifacts for the patch with the given number are stored.
+    fn patch_dir(&self, patch_number: usize) -> PathBuf {
+        self.patches_dir().join(patch_number.to_string())
+    }
+
+    /// The path to the runnable patch artifact with the given number. Runnable patch artifact files are
+    /// named <patch_number>.vmcode
+    fn patch_artifact_path(&self, patch_number: usize) -> PathBuf {
+        self.patch_dir(patch_number).join(PATCH_ARTIFACT_FILENAME)
+    }
+
+    fn patch_info_for_number(&self, patch_number: usize) -> PatchInfo {
+        PatchInfo {
+            path: self.patch_artifact_path(patch_number),
+            number: patch_number,
+        }
+    }
+
+    /// Checks that the patch with the given number:
+    ///   - Has an artifact on disk
+    ///   - That artifact on disk is the same size it was when it was installed
+    ///
+    /// Returns Ok if the patch is bootable, or an error if it is not.
+    fn validate_patch_is_bootable(&self, patch: &PatchMetadata) -> Result<()> {
+        let artifact_path = self.patch_artifact_path(patch.number);
+        if !Path::exists(&artifact_path) {
+            bail!(
+                "Patch {} does not exist at {}",
+                patch.number,
+                artifact_path.display()
+            );
+        }
+
+        let artifact_size_on_disk = std::fs::metadata(&artifact_path)?.len();
+        if artifact_size_on_disk != patch.size {
```

Comment:
Makes me wonder if we should report the boot size as part of our patch_available callback to the server, or some other mechanism for possibly detecting tampering in the field.

Reply:
(Not something to do in this patch, more just seeding your thoughts.)